tests/backbeat/tests/retry/retry.js

```diff
@@ -0,0 +1,164 @@
+                    return next();
+                });
+            }),
+        next => times(N, (n, cb) =>
+            scalityUtils.waitWhileFailedCRR(srcBucket, keys[n], cb),
+        next),
+        next => timesSeries(N, (n, cb) =>
+            scalityUtils.compareObjectsAWS(srcBucket, destFailBucket,
+                keys[n], undefined, cb),
+        next),
+    ], done);
+}
+
+describe('Backbeat replication retry', function() {
+    this.timeout(REPLICATION_TIMEOUT);
+    const roleArn = 'arn:aws:iam::root:role/s3-replication-role';
+
+    beforeEach(done => series([
+        next => scalityUtils.createVersionedBucket(srcBucket, next),
+        next => scalityUtils.putBucketReplicationMultipleBackend(srcBucket,
+            destFailBucket, roleArn, destFailLocation, next),
+    ], done));
+
+    afterEach(done => parallel([
+        next => scalityUtils.deleteVersionedBucket(srcBucket, next),
+        next => awsUtils.deleteAllVersions(destFailBucket,
+            `${srcBucket}/${keyPrefix}`, next),
+    ], done));
+
+    [1, 2, 128].forEach(N => {
+        it(`should retry ${N} failed object(s)`, done => {
+            let versionId;
+            let postBody;
+            const keys = [];
+            for (let i = 0; i < N; i++) {
+                keys.push(`${key}-${i}`);
+            }
+            return performRetries(N, keys, done);
+        });
+    });
+
+    it('should get correct CRR metrics when a retry occurs', function(done) {
+        this.retries(2); // Test is dependent on metrics not expiring.
+        const path = `/_/backbeat/api/metrics/crr/${destFailLocation}`;
+        let prevBacklog;
+        let prevCompletions;
+        let prevFailures;
+        return series([
+            next => makeGETRequest(path, (err, res) => {
+                assert.ifError(err);
+                assert.equal(res.statusCode, 200);
+                getResponseBody(res, (err, body) => {
+                    assert.ifError(err);
+                    prevBacklog = body.backlog.results;
+                    prevCompletions = body.completions.results;
+                    prevFailures = body.failures.results;
+                    return next();
+                });
+            }),
+            next => performRetries(1, [key], next),
+            next => {
+                let shouldContinue = false;
+                return doWhilst(callback =>
+                    makeGETRequest(path, (err, res) => {
+                        assert.ifError(err);
+                        assert.strictEqual(res.statusCode, 200);
+                        getResponseBody(res, (err, body) => {
+                            assert.ifError(err);
+                            const { results } = body.completions;
+                            // We have reached an expiration of metrics.
+                            if (results.count < prevCompletions.count) {
+                                return callback(new Error('Invalid test case'));
```

Comment:
I'm not sure but a bug in the code may well trigger this case (any bug in metrics, say, integer conversion error etc.), so instead maybe use a regular `assert`?

Reply:
Very good point. Updated.